src/os/FileStore.cc

```diff
@@ -2930,7 +2930,8 @@ int FileStore::read(
     memset(&st, 0, sizeof(struct stat));
     int r = ::fstat(**fd, &st);
     assert(r == 0);
-    len = st.st_size;
+    assert(offset < st.st_size);
+    len = st.st_size - offset;
```

Comment:
There shouldn't be an assert here when the offset, len are coming from a client.  A truncation could race with the client request, so any offset > 0 could become problematic.  Also, if the file is empty (st.size == 0) then the offset=0, len=0 should be the same as offset=0, len=1 for an empty file (which returns no data from read).

So the solution is if offset >= st.st_size just return no data otherwise the len = st.st_size - offset is fine.